32blit-stm32/Src/32blit/32blit_battery.cpp

```diff
@@ -0,0 +1,129 @@
+// 32blit_battery.cpp
+//
+// 32Blit Firmware Battery features
+//
+// 32Blit uses a BQ24295 to control battery/charging/etc. Use below programming guide for I2C status description/interpretation
+//
+// > Product page: https://www.ti.com/product/BQ24295
+// > Programming Guide: https://www.ti.com/lit/ds/symlink/bq24295.pdf?ts=1609111702094&ref_url=https%253A%252F%252Fwww.ti.com%252Fproduct%252FBQ24295
+//
+
+#include "32blit.h"
+#include "32blit_battery.h"
+
+using namespace blit;
+
+//
+// Return information about the battery
+//
+BatteryInformation BatteryState::get_battery_info() {
+  BatteryInformation info;
+
+  info.charge_status = get_battery_charge_status();
+  info.charge_text = get_charge_status_string();
+
+  info.vbus_status = get_battery_vbus_status();
+  info.vbus_text = get_vbus_status_string();
+
+  info.voltage = battery;
+
+  return info;
+}
+
+//
+// Update battery status and fault (read using I2C)
+//
+void BatteryState::update_battery_status ( uint8_t status, uint8_t fault ) {
+  battery_status = status;
+  battery_fault = fault;
+}
+
+//
+// Update the charge value
+//
+void BatteryState::update_battery_charge ( float charge_value ) {
+  battery_average.add(charge_value);
+  battery = battery_average.average();
+}
+
+//
+// Return the current battery status
+//
+BatteryChargeStatus BatteryState::get_battery_charge_status() {
+  auto status = (battery_status >> 4) & 0b11;
+  switch(status){
+    case 0b00: // Not charging
+      return BatteryChargeStatus::NotCharging;
+    case 0b01: // Pre-charge
+      return BatteryChargeStatus::PreCharging;
+    case 0b10: // Fast Charging
+      return BatteryChargeStatus::FastCharging;
+    case 0b11: // Charge Done
+      return BatteryChargeStatus::ChargingComplete;
```

Comment:
Hmm, I'd define the enum values at these to avoid translating. Then the string getters could also use the enums for even less magic numbers.

Reply:
+1 to reducing magic number count.